### ISSUE TITLE
How to implement new weapons

### ISSUE THREAD

**Issue by flip40** [DemoProductions/shmup#90]
If we want to add features to weapons, should we either:

**A**: Use separate scripts for each weapon type (using inheritance to avoid copy-pasta).

Pros
- No editor shenanigans required.

Cons
- We will have multiple different Weapon scripts lying around instead of a single one.
- ~~GetComponent<SuperClass> does **NOT** work. Considering we use Weapons as prefabs, this may not be an issue, but if we ever need to use that function, it will not work in the abstract.~~

**B**: Use a single script with options (similar to how EnemyMovement works).

Pros
- Single script for defining weapon behavior.
- ~~No issues with GetComponent<SuperClass> due to being contained in a single script.~~

Cons
- Will likely require lots of editor shenanigans (code) to best function in a way that makes sense (EnemyMovement as example).

**Comment by ghost** [comment]
Both options seems reasonable at the moment, but in the longer term option A sounds more convenient. If we start having a larger list of weapons, I imagine the Weapon script would get convoluted and so would the editor configurations. The GetComponent<SuperClass> con may not be too bad, we would have our weapon scripts named appropriately (ex: [replaceWithWeaponName]Weapon). I think the same for EnemyMovement, except it seems more difficult to separate out chunks of code for different movement types. Maybe we can though, the EnemyMovement class we have now can be a base, but say we want certain enemies to do so something like teleport or move in some unique shape.

**Comment by flip40** [comment]
Yeah, lets do A. I was concerned for a second as I had just realized I don't actually know how we would get a weapon by GetComponent without using the super class... And according to Google, wherever it was that I heard that you couldn't GetComponent<SuperClass> is completely wrong! So A is in fact the superior option, and GetComponent<Weapon> will work as expected. I updated the original post to cross off the mistaken pros / cons in regards to that.
